### ISSUE TITLE
have diagm construct a Diagonal matrix?

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#10105]
This would allow the linear algebra way of writing lots of idioms efficient.

**Comment by andreasnoack** [comment]
Yes. In many cases people would be better off with a `Diagonal` matrix. However, we need to consider the cases where people would like to manipulate the result with `setindex!`, e.g. if they want to construct the matrix
```julia
3x3 Array{Int64,2}:
 1  0  1
 0  2  0
 0  0  3
```
The `setindex` situation for the special matrices is far from settled. It is unclear if we it is desirable to support all the kinds of indexing that `Array`s support. As an example, `XTriangular` now throws a runtime error when you try to modify an element in the zero region. We do so for type stability. The other type stable solution would be to always promote to `Matrix`.

So in short, should we require that the user calls `full(D)` before modification or should `setindex!` promote?

Another issue is the naming. Why use `diagm` instead of the constructor `Diagonal`?

**Comment by nalimilan** [comment]
> Another issue is the naming. Why use diagm instead of the constructor Diagonal?

Maybe precisely this difference can be used to make it easy to get either a standard `Array` (with `diagm`) or a `Diagonal` matrix (with `Diagonal`) depending on your needs.

**Comment by ViralBShah** [comment]
What you want is `diagm` to give you a regular dense diagonal matrix, and use `Diagonal` if you want the storage optimized version, which is not completely general purpose. Otherwise, people will have to spray `full()` all over their codes.

**Comment by StefanKarpinski** [comment]
Ok, fair enough.
